ASMDocumentation.md

```diff
@@ -0,0 +1,5841 @@
+-        mov     DWORD PTR [esp-8], eax
+-        fld     DWORD PTR [esp-8]
+-        fmul    ST(0), ST(0)
+-        fld     DWORD PTR [esp-8]
+-        fxch    ST(1)
+-        fmul    DWORD PTR [esp+4]
+-        fld     DWORD PTR [esp-12]
+-        fld     ST(0)
+-        fsub    ST(0), ST(2)
+-        fld     ST(1)
+-        fxch    ST(1)
+-        fmul    ST(3), ST(0)
+-        fmul    ST(3), ST(0)
+-        fmulp   ST(4), ST(0)
+-        fsub    ST(0), ST(2)
+-        fmul    ST(2), ST(0)
+-        fmul    ST(3), ST(0)
+-        fmulp   ST(2), ST(0)
+-        fxch    ST(1)
+-        fsubp   ST(1), ST(0)
+-        fmulp   ST(1), ST(0)
++        push    ebp
++        mov     ebp, esp
++        sub     esp, 20                             ; 00000014H
++        mov     DWORD PTR _this$[ebp], ecx
++        movss   xmm0, DWORD PTR _a$[ebp]
++        ucomiss xmm0, DWORD PTR __real@00000000
++        lahf
++        test    ah, 68                              ; 00000044H
++        jp      SHORT $LN2@Inv_Sqrt
++        fldz
++        jmp     SHORT $LN1@Inv_Sqrt
++$LN2@Inv_Sqrt:
++        movss   xmm0, DWORD PTR __real@3fc00000
++        movss   DWORD PTR _threehalfs$[ebp], xmm0
++        movss   xmm0, DWORD PTR _a$[ebp]
++        mulss   xmm0, DWORD PTR __real@3f000000
++        movss   DWORD PTR _x2$[ebp], xmm0
++        movss   xmm0, DWORD PTR _a$[ebp]
++        movss   DWORD PTR _y$[ebp], xmm0
++        mov     eax, DWORD PTR _y$[ebp]
++        mov     DWORD PTR _i$[ebp], eax
++        mov     ecx, DWORD PTR _i$[ebp]
++        sar     ecx, 1
++        mov     edx, 1597463007                     ; 5f3759dfH
++        sub     edx, ecx
++        mov     DWORD PTR _i$[ebp], edx
++        movss   xmm0, DWORD PTR _i$[ebp]
++        movss   DWORD PTR _y$[ebp], xmm0
++        movss   xmm0, DWORD PTR _x2$[ebp]
++        mulss   xmm0, DWORD PTR _y$[ebp]
++        mulss   xmm0, DWORD PTR _y$[ebp]
++        movss   xmm1, DWORD PTR __real@3fc00000
++        subss   xmm1, xmm0
++        mulss   xmm1, DWORD PTR _y$[ebp]
++        movss   DWORD PTR _y$[ebp], xmm1
++        fld     DWORD PTR _y$[ebp]
++$LN1@Inv_Sqrt:
++        mov     esp, ebp
++        pop     ebp
+         ret     4
+ float WWMath::Inv_Sqrt(float) ENDP                      ; WWMath::Inv_Sqrt
+```
+
+</details>
+
+---
+
+<details>
+<summary>Generals/Code/Libraries/Source/WWVegas/WWDebug/wwprofile.cpp</summary>
+
+This file includes the following assembly block:
+
+```c++
+inline void WWProfile_Get_Ticks(_int64 * ticks)
+{
+#ifdef _UNIX
+	*ticks = 0;
+#else 
+	__asm
+	{
+		push edx;
+		push ecx;
+		mov ecx,ticks;
+		_emit 0Fh
+		_emit 31h
+		mov [ecx],eax;
+		mov [ecx+4],edx;
+		pop ecx;
+		pop edx;
+	}
+#endif
+}
+```
+
+This assembly block is trying to get the current ticks using
+the [RDTSC](https://www.aldeid.com/wiki/X86-assembly/Instructions/rdtsc)
+instruction.
+
+My goto equivalent is:
+
+```c++
+#include <intrin.h>
+
+inline void WWProfile_Get_Ticks(_int64 * ticks)
+{
+#ifdef _UNIX
```

Comment:
GCC/clang have __rdtsc intrinsic as well. Ideally we implement something for other CPUs too.